classes/user_enrolment_repository.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace local_augmented_teacher;
+
+class user_enrolment_repository
+{
+
+    private \moodle_database $db;
+
+    public function __construct(\moodle_database $DB) {
+        $this->db = $DB;
+    }
+
+    /**
+     * Get user enrolments by course ids
+     * @param array $course_ids
+     * @return array
+     * @throws \dml_exception
+     */
+    public function get_by_course_ids(array $course_ids): array {
+        [$in_sql, $params] = $this->db->get_in_or_equal($course_ids);
+
+        $sql = 'SELECT id, status, userid
+                FROM {user_enrolments}
+                WHERE enrolid IN (SELECT id
+                                  FROM {enrol}
+                                  WHERE courseid ' . $in_sql . ')';
+
+        return $this->db->get_records_sql($sql, $params);
+    }
+
+    /**
+     * Get user enrolments by reminders
+     * @param $reminders
+     * @return array
+     * @throws \coding_exception
+     * @throws \dml_exception
+     */
+    public function get_from_reminders(\moodle_recordset $reminders): array {
+        $course_ids = [];
+
+        foreach ($reminders as $reminder) {
+            if ($cm = get_coursemodule_from_id('', $reminder->cmid)) {
```

Comment:
This might be a performance issue. It's not cached and it's accessing the DB in every loop. It's a Moodle core function but it's quite simple SQL. You need to preload the result set like you did with the reminders and user enrolments.